### PR TITLE
Implement working v1

### PR DESCRIPTION
**Pull request by thusoy**
This implements a working Cloudflare to Honeycomb log forwarder. It supports sampling by status code, enriching based on url pattern matching, query parameter extraction with a safelist, and loading honeycomb API keys from vault.

When run against a 30MB test file with data (~300k entries) and a sampling rate for 200s of 10 (~39k total processed events) it runs in ~45s, which has plenty of margin on the 540s timeout. The higher the sampling rate the faster it'll run since it'll skip the enrichment on the skipped events.

This also uses a separate GCS bucket for locking since Cloud Functions are at-least-once to prevent duplicate submission, and stores already processed files in a separate location. Combined this ensures that two functions started at the same time will not double process a file, and if a function is retried later it also won't reprocess. However, if a function crashes mid-way during processing there is a recovery mechanism that detects stale locks and deletes them after 1 day to allow re-processing on failure.

### CODE REVIEW

File: honeyflare/__init__.py [hunk 1]
```diff
@@ -1,2 +1,197 @@
...
+
+        local_path = download_file(bucket, object_name)
+
+        for sample_rate, entry in get_sampled_file_entries(local_path, sampling_rate_by_status):
+            event = libhoney_client.new_event()
+            event.sample_rate = sample_rate
+            enrich_entry(entry, compiled_patterns, query_param_filter)
+            event.add(entry)
+            event.created_at = datetime.datetime.utcfromtimestamp(entry['EdgeEndTimestamp']/1e9)
+            event.send_presampled()
```
Comment (dextersd) [inline honeyflare/__init__.py]: As we discussed not super familiar with events vs trace/spans but I'm guessing this is to not double sample (as they are already filtered at `get_sampled_file_entries`)

File: honeyflare/__init__.py [hunk 2]
```diff
@@ -1,2 +1,197 @@
...
+    try:
+        blob.upload_from_string(b'')
+    except PreconditionFailed:
+        # Another invocation has already processed this but it failed to be
+        # caught in the lock. Ignore
+        pass
+
+
+def _processed_blob(bucket, object_name):
+    return bucket.blob('completed/%s' % object_name)
```
Comment (dextersd) [inline honeyflare/__init__.py]: I'm guessing this is creating a file in the bucket in the completed directory, would it make sense to use something more ephemeral like a redis key expiring after 24 hours?

Comment (dextersd) [reply honeyflare/__init__.py]: I guess keeping everything scoped to gcloud makes it easier at not having to worry about dependencies

File: honeyflare/vault.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    key = parsed_parameters['key'][0]
+
+    # Clean up to not affect other https queries
+    del os.environ['REQUESTS_CA_BUNDLE']
+
+    return honeycomb_key['data']['data'][key]
+
+
+def get_auth_jwt(vault_role):
+    url = 'http://metadata/computeMetadata/v1/instance/service-accounts/default/identity'
```
Comment (dextersd) [inline honeyflare/vault.py]: This url structure is specifically for vault?

File: main.py [hunk 4]
```diff
@@ -1,15 +1,106 @@
...
+# Check for required envvars to fail early on invalid deployments
+honeycomb_dataset = os.environ.get('HONEYCOMB_DATASET')
+if honeycomb_dataset is None:
+    raise ValueError('Missing environment variable HONEYCOMB_DATASET')
+
+honeycomb_meta_dataset = os.environ.get('HONEYCOMB_META_DATASET')
+if honeycomb_meta_dataset is None:
+    raise ValueError('Missing environment variable HONEYCOMB_META_DATASET')
+
+honeycomb_key = os.environ.get('HONEYCOMB_KEY')
```
Comment (dextersd) [inline main.py]: How are these env variables passed in? Is there any reason we wouldn't use vault to also distribute things like `HONEYCOMB_KEY`?

File: honeyflare/__init__.py (hunk 1 above)
Comment (thusoy) [reply honeyflare/__init__.py]: Correct.

File: honeyflare/__init__.py (hunk 2 above)
Comment (thusoy) [reply honeyflare/__init__.py]: Yup, that was the idea, no extra services managed by us.

File: honeyflare/vault.py (hunk 3 above)
Comment (thusoy) [reply honeyflare/vault.py]: No, this is the GCP metadata service. It's somewhat similar for all cloud providers, but the specifics are usually provider dependent. The only thing vault-specific here is the `audience` tag we request, which must be `vault/<role>`.

File: main.py (hunk 4 above)
Comment (thusoy) [reply main.py]: They are deployed with the function, in our case defined in the terraform module. Secrets stores are best used for secrets only, using them for other things makes it much harder to just inspect what values are set for general config.

**Pull request merged**